src/include/execution/vm/bytecode_handlers.h

```diff
@@ -1321,215 +1323,202 @@ VM_OP void OpIndexIteratorPerformInit(terrier::execution::sql::IndexIterator *it
+    /* Read */                                                                                             \
+    bool null = false;                                                                                     \
+    auto *ptr = pr->Get<CppType, true>(col_idx, &null);                                                    \
+    /* Set */                                                                                              \
+    out->is_null_ = null;                                                                                  \
+    out->val_ = null ? 0 : *ptr;                                                                           \
+  }
 
-VM_OP_HOT void OpIndexIteratorGetDecimal(terrier::execution::sql::Decimal *out,
-                                         UNUSED_ATTRIBUTE terrier::execution::sql::IndexIterator *iter,
-                                         UNUSED_ATTRIBUTE uint16_t col_idx) {
-  // Set
-  out->is_null_ = false;
-  out->val_ = 0;
+GEN_PR_SCALAR_SET_CALLS(TinyInt, Integer, int8_t);
+GEN_PR_SCALAR_SET_CALLS(SmallInt, Integer, int16_t);
+GEN_PR_SCALAR_SET_CALLS(Int, Integer, int32_t);
+GEN_PR_SCALAR_SET_CALLS(BigInt, Integer, int64_t);
+GEN_PR_SCALAR_SET_CALLS(Real, Real, float);
+GEN_PR_SCALAR_SET_CALLS(Double, Real, double);
+GEN_PR_SCALAR_GET_CALLS(TinyInt, Integer, int8_t);
+GEN_PR_SCALAR_GET_CALLS(SmallInt, Integer, int16_t);
+GEN_PR_SCALAR_GET_CALLS(Int, Integer, int32_t);
+GEN_PR_SCALAR_GET_CALLS(BigInt, Integer, int64_t);
+GEN_PR_SCALAR_GET_CALLS(Real, Real, float);
+GEN_PR_SCALAR_GET_CALLS(Double, Real, double);
+
+VM_OP_HOT void OpPRSetVarlen(terrier::execution::sql::ProjectedRowWrapper *pr, uint16_t col_idx,
+                             terrier::execution::sql::StringVal *val) {
+  if (val->len_ <= terrier::storage::VarlenEntry::InlineThreshold()) {
+    terrier::storage::VarlenEntry varlen(terrier::storage::VarlenEntry::CreateInline(
+        reinterpret_cast<const terrier::byte *>(val->Content()), val->len_));
+    pr->Set<terrier::storage::VarlenEntry, false>(col_idx, varlen, val->is_null_);
+  } else {
+    terrier::storage::VarlenEntry varlen(terrier::storage::VarlenEntry::Create(
+        reinterpret_cast<const terrier::byte *>(val->Content()), val->len_, false));
+    pr->Set<terrier::storage::VarlenEntry, false>(col_idx, varlen, val->is_null_);
+  }
 }
 
-VM_OP_HOT void OpIndexIteratorGetTinyIntNull(terrier::execution::sql::Integer *out,
-                                             terrier::execution::sql::IndexIterator *iter, uint16_t col_idx) {
-  // Read
-  bool null = false;
-  auto *ptr = iter->Get<int8_t, true>(col_idx, &null);
-
-  // Set
-  out->is_null_ = null;
-  out->val_ = null ? 0 : *ptr;
+VM_OP_HOT void OpPRSetVarlenNull(terrier::execution::sql::ProjectedRowWrapper *pr, uint16_t col_idx,
+                                 terrier::execution::sql::StringVal *val) {
+  if (val->len_ <= terrier::storage::VarlenEntry::InlineThreshold()) {
```

Comment:
You need to move this logic out of the handlers into `StringVal`. It doesn't belong there either, but it's better than here.